File: ds/stack.py
# Implementation of stack data type
# Stack supports LIFO - last in first out structure

class Stack:

    def __init__(self):
        self.stack = []

    # function to push element into the stack
    def push(self, data):
        self.stack.append(data)

    # function to pop element from the stack
    def pop(self):
        data = self.stack[-1]
        del self.stack[-1]
        return data

    # function to peek the element of the stack
    def peek(self):
        return self.stack[-1]

    # function to check if the stack is empty
    def isEmpty(self):
        return self.stack == []

    # function to get the size of the stack
    def size(self):
        return len(self.stack)

    # function to print all the elements of a stack
    def printStack(self):
        print("=======")

        for item in self.stack:
            print(item)

        print("=======")


# Creating and testing values of the stack
stack = Stack()

stack.push(10)
stack.push(12)

stack.printStack()

print(stack.peek())
print(stack.pop())
print(stack.peek())

stack.printStack()

File: ds/Stack.cpp
#include <iostream>
using namespace std;

class Stack
{
public:
    int *arr;
    int size;
    int top;

    Stack(int size)
    {
        this->size = size;
        arr = new int[size];
        top = -1;
    }

    /**
     * pushing into the stack
     */
    void push(int element)
    {
        if (size - top > 0)
        {
            top++;
            arr[top] = element;
        }
        else
        {
            cout << "Stack overflow" << endl;
        }
    }

    /**
     * Popping from the stack
     */
    void pop()
    {
        if (top >= 0)
        {
            top--;
        }
        else
        {
            cout << "Stack undeflow" << endl;
        }
    }

    /**
     * Peeking into the stack
     */
    int peek()
    {
        if (isEmpty())
        {
            return -1;
        }
        else
        {
            return arr[top];
        }
    }

    bool isEmpty()
    {
        if (top == -1)
        {
            return true;
        }
        else
        {
            return false;
        }
    }
};

// Driver code
int main()
{
    Stack st(10);
    st.push(1);
    st.push(2);
    st.push(3);

    cout << st.peek() << endl;

    st.pop();

    cout << st.peek() << endl;

    return 0;
}